contracts/Governance/Timelock.sol

```diff
@@ -1,123 +1,177 @@
+    }
 
-        bytes memory callData;
+    function approveTransactionsBatch(bytes32 txParamsHash) external onlyOwner {
+        _approveTransactionsBatchInternal(txParamsHash);
+    }
 
-        if (bytes(signature).length == 0) {
-            callData = data;
-        } else {
-            callData = abi.encodePacked(bytes4(keccak256(bytes(signature))), data);
-        }
+    function executeTransactionsBatch(Transaction[] memory transactions, uint256 executionStartTimestamp) external payable onlyOwner {
+        _executeTransactionsBatchInternal(transactions, executionStartTimestamp);
+    }
+
+    function approveAndExecuteTransactionsBatch(Transaction[] memory transactions, uint256 executionStartTimestamp) external payable onlyOwner {
+        bytes32 txParamsHash = keccak256(abi.encode(transactions, executionStartTimestamp));
+
+        _approveTransactionsBatchInternal(txParamsHash);
+        _executeTransactionsBatchInternal(transactions, executionStartTimestamp);
+    }
+
+    function approveAndExecuteTransactionsBatchRaw(bytes calldata txParamsData) external payable onlyOwner {
+        bytes32 txParamsHash = keccak256(txParamsData);
+        (Transaction[] memory transactions, uint256 executionStartTimestamp) = abi.decode(txParamsData, (Transaction[], uint256));
+
+        _approveTransactionsBatchInternal(txParamsHash);
+        _executeTransactionsBatchInternal(transactions, executionStartTimestamp);
+    }
+
+    function _approveTransactionsBatchInternal(bytes32 txParamsHash) internal onlyOwner {
+        require(queuedTransactions[txParamsHash] == TransactionStatus.Queued, "Timelock: transaction is not queued");
+
+        queuedTransactions[txParamsHash] = TransactionStatus.Approved;
 
-        // Execute the call
-        (bool success, bytes memory returnData) = target.call{value: value}(callData);
-        require(success, "Timelock::executeTransaction: Transaction execution reverted.");
+        emit ApprovedTransactionsBatch(txParamsHash);
+    }
+
+    function _executeTransactionsBatchInternal(Transaction[] memory transactions, uint256 executionStartTimestamp) internal {
+        bytes32 txParamsHash = keccak256(abi.encode(transactions, executionStartTimestamp));
+
+        require(queuedTransactions[txParamsHash] == TransactionStatus.Approved, "Timelock: Transaction hasn't been approved.");
+        require(block.timestamp >= executionStartTimestamp, "Timelock: Transaction hasn't surpassed the execution delay.");
+        require(block.timestamp <= executionStartTimestamp + executionGracePeriod, "Timelock: Transaction is stale.");
 
-        emit ExecuteTransaction(txHash, target, value, signature, data, eta);
+        delete queuedTransactions[txParamsHash];
 
-        return returnData;
+        for (uint256 i = 0; i < transactions.length; i++) {
+            (
+                bool success, /* ignore the rest */
+
+            ) = transactions[i].recipient.call{value: transactions[i].value}(transactions[i].data);
```

Comment:
Thinking out loud here about security - we followed the [Checks Effects Interactions](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html) pattern, right?

Seems like it to me, but wanted to see what you guys think.

Reply:
Looks like we did :)